include/exadg/structure/spatial_discretization/operators/nonlinear_operator.cpp

```diff
@@ -275,27 +282,83 @@ template<int dim, typename Number>
+       boundary_type == BoundaryType::RobinSpringDashpotPressure)
     {
-      tensor F = get_F<dim, Number>(integrator.get_gradient(q));
-      vector N = integrator.get_normal_vector(q);
-      // da/dA * n = det F F^{-T} * N := n_star
-      // -> da/dA = n_star.norm()
-      vector n_star = determinant(F) * transpose(invert(F)) * N;
-      // t_0 = da/dA * t
-      traction *= n_star.norm();
+      if(operator_type == OperatorType::inhomogeneous or operator_type == OperatorType::full)
+      {
+        traction -= calculate_neumann_value<dim, Number>(
+          q, integrator, boundary_type, boundary_id, this->operator_data.bc, this->time);
+
+        if(this->operator_data.pull_back_traction)
+        {
+          tensor F = get_F<dim, Number>(integrator.get_gradient(q));
+          vector N = integrator.get_normal_vector(q);
+          // da/dA * n = det F F^{-T} * N := n_star
+          // -> da/dA = n_star.norm()
+          vector n_star = determinant(F) * transpose(invert(F)) * N;
+          // t_0 = da/dA * t
+          traction *= n_star.norm();
+        }
+      }
+    }
+
+    // check boundary ID in robin_k_c_p_param to add boundary mass integrals from Robin boundaries
+    // on BoundaryType::NeumannCached or BoundaryType::RobinSpringDashpotPressure
+    if(boundary_type == BoundaryType::NeumannCached or
+       boundary_type == BoundaryType::RobinSpringDashpotPressure)
+    {
+      if(operator_type == OperatorType::homogeneous or operator_type == OperatorType::full)
+      {
+        auto const it = this->operator_data.bc->robin_k_c_p_param.find(boundary_id);
+
+        if(it != this->operator_data.bc->robin_k_c_p_param.end())
+        {
+          bool const   normal_projection_displacement = it->second.first[0];
+          double const coefficient_displacement       = it->second.second[0];
+
+          if(normal_projection_displacement)
+          {
+            vector const N = integrator.get_normal_vector(q);
+            traction += N * (coefficient_displacement * (N * integrator.get_value(q)));
+          }
+          else
+          {
+            traction += coefficient_displacement * integrator.get_value(q);
+          }
+
+          if(this->operator_data.unsteady)
+          {
+            bool const   normal_projection_velocity = it->second.first[1];
+            double const coefficient_velocity       = it->second.second[1];
+
+            if(normal_projection_velocity)
+            {
+              vector const N = integrator.get_normal_vector(q);
+              traction += N * (coefficient_velocity * this->scaling_factor_mass_boundary *
+                               (N * integrator.get_value(q)));
+            }
+            else
+            {
+              traction +=
+                coefficient_velocity * this->scaling_factor_mass_boundary * integrator.get_value(q);
+            }
+          }
+        }
+      }
```

Comment:
This code looks like a duplication of the one above in `linear_operator.cpp`. Is there any difference between these two version?

In the previous code version, one could say we had a function call `calculate_neumann_value()` in both cases (linear operator and nonlinear operator), but here we have a sequence of instructions and calculations that are duplicated, which should definitely be avoided.